contracts/andromeda_astroport/src/contract.rs

```diff
@@ -98,6 +123,180 @@ pub fn execute(
+        Recipient::Addr(info.sender.to_string()),
+    )?;
+
+    // In the case where we want to witdraw the LP token.
+    contract.add_withdrawable_token(
+        deps.storage,
+        &pair.liquidity_token.to_string(),
+        &AssetInfo::Cw20(pair.liquidity_token),
+    )?;
+    let mut initial_transfer_messages: Vec<SubMsg> = vec![];
+    for asset in assets.iter() {
+        if let AssetInfo::Cw20(contract_addr) = &asset.info {
+            initial_transfer_messages.push(SubMsg::new(WasmMsg::Execute {
+                contract_addr: contract_addr.to_string(),
+                // User needs to allow this contract to transfer tokens.
+                msg: encode_binary(&Cw20ExecuteMsg::TransferFrom {
+                    owner: info.sender.to_string(),
+                    recipient: env.contract.address.to_string(),
+                    // The excess is already deducted from this amount.
+                    amount: asset.amount,
+                })?,
+                funds: vec![],
+            }));
+            initial_transfer_messages.push(SubMsg::new(WasmMsg::Execute {
+                contract_addr: contract_addr.to_string(),
+                // Allow the pair address to transfer from here.
+                msg: encode_binary(&Cw20ExecuteMsg::IncreaseAllowance {
+                    spender: pair.contract_addr.to_string(),
+                    // The excess is already deducted from this amount.
+                    amount: asset.amount,
+                    expires: None,
+                })?,
+                funds: vec![],
+            }));
+        }
+    }
+
+    let native_funds = get_native_funds_from_assets(&assets);
+    Ok(Response::new()
+        .add_submessages(initial_transfer_messages)
+        .add_submessages(overflow_messages)
+        .add_message(CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: pair.contract_addr.to_string(),
+            msg: encode_binary(&AstroportPairExecuteMsg::ProvideLiquidity {
+                assets: assets.map(|a| a.into()),
+                slippage_tolerance,
+                auto_stake,
+                // Not strictly neccessary but to be explicit.
+                receiver: Some(env.contract.address.to_string()),
+            })?,
+            funds: native_funds,
+        }))
+        .add_attribute("action", "provide_liquidity"))
+}
+
+fn execute_withdraw_liquidity(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    pair_address: String,
+    amount: Option<Uint128>,
+    recipient: Option<Recipient>,
+) -> Result<Response, ContractError> {
+    let sender = info.sender.as_str();
+    let contract = ADOContract::default();
+    require(
+        contract.is_owner_or_operator(deps.storage, sender)?,
+        ContractError::Unauthorized {},
+    )?;
+    let recipient = recipient.unwrap_or_else(|| Recipient::Addr(sender.to_owned()));
+    let pair_info = query_pair_given_address(&deps.querier, pair_address)?;
+    let total_amount = query_token_balance(
```

Comment:
I assume this amount is validated in the Astroport contract for insufficient funds?

Reply:
This represents how many LP tokens our contract holds which I don't think needs to be validated anywhere, since for those tokens to exist there must be the related amount of underlying tokens given for liquidity.